examples/NAT.cs

```diff
@@ -36,223 +36,348 @@
-      to.tcp_port = p_tcp.DestinationPort;
-
-      // Generate data for the mapping
-      lock(my_address)
-      {
-        // NOTE can't use Interlocked.Increment since the type of next_port is short not int.
-        to.assigned_tcp_port = next_port++;
-      }
+      // If a TCP SYN, then add a mapping for the new connection.
+      natPort = GetNATPort();
+      // Add to NAT_MapToOutside
+      NAT_MapToOutside[out_key] = natPort;
+      // Add to NAT_MapToInside
+      var in_key = new MapToInside_Key(p_ip.DestinationAddress, p_tcp.DestinationPort, natPort);
+      var internalNode = new InternalNode(p_ip.SourceAddress, p_tcp.SourcePort, incomingPort, p_eth.SourceHwAddress);
+      NAT_MapToInside[in_key] = internalNode;
+      Console.WriteLine("Added mapping");
+    }
+    else if (!NAT_MapToOutside.TryGetValue(out_key, out natPort))
+    {
+      // Not a SYN, and no existing connection, so drop.
+      return Port_Drop;
+    }
 
-      if (!port_mapping.TryAdd(to, from))
-      {
-        Console.WriteLine("Concurrent update of port_mapping[to] where 'to'={0}", to);
-      }
-      if (!port_reverse_mapping.TryAdd(from, to))
-      {
-        Console.WriteLine("Concurrent update of port_reverse_mapping[from] where 'from'={0}", from);
-      }
+    // Rewrite the packet.
+    p_tcp.SourcePort = natPort;
+    p_ip.SourceAddress = my_address;
+    // Update checksums. NOTE IP updated before TCP
+    ((IPv4Packet)p_ip).UpdateIPChecksum();
+    p_tcp.UpdateTCPChecksum();
+    // Update destination MAC address
+    p_eth.DestinationHwAddress = next_outside_hop_mac;
+    return Port_Outside;
+  }
 
-      // Rewrite the packet.
-      p_tcp.SourcePort = to.assigned_tcp_port.Value;
-      p_ip.SourceAddress = my_address;
-      // Update checksums.
-      p_tcp.UpdateTCPChecksum();
-      ((IPv4Packet)p_ip).UpdateIPChecksum();
-      return Port_Outside;
-    } else {
-      // Not a SYN, so this must be part of an ongoing connection.
-      if (!port_reverse_mapping.TryGetValue(from, out to))
-      {
-        // if we don't have a mapping for it then drop.
-        return Port_Drop;
-      } else {
-        //  otherwise apply the mapping (replacing source IP address and TCP port) and forward to the Outside port.
-        p_tcp.SourcePort = to.assigned_tcp_port.Value;
-        p_ip.SourceAddress = my_address;
-        // Update checksums.
-        p_tcp.UpdateTCPChecksum();
-        ((IPv4Packet)p_ip).UpdateIPChecksum();
-        return Port_Outside;
-      }
+  private ushort GetNATPort()
+  {
+    lock(portLock)
+    {
+      // NOTE can't use Interlocked.Increment since the type of next_port is short not int.
+      return next_port++;
     }
   }
 
+  sealed class InternalNode : IEquatable<InternalNode>
+  {
+    private readonly IPAddress _Address; // FIXME we should really use immutable values since we are using a dictionary
+    private readonly ushort _Port;
+    private readonly int _NetworkPort;
```

Comment:
I can intuit what's intended to be stored in _Port and how it differs from _NetworkPort (partly from the types of both variables), but owing to the similarity of their names it would be useful to have a comment that states what each of the two is intended to store.